RedundantTypeAnnotationRule: Rewrite with SwiftSyntax and allow ignoring certain attributes

tonell-m opened the pull request:
Hi, first time contributing here so just let me know if something is wrong or missing in my PR 🙂 

This PR resolves #5366 by adding a new `ignored_annotations` configuration to this rule that accepts a set of strings listing annotations that should disable this rule for properties that are marked with it. Currently this is already done arbitrarily for `@IBInspectable` but now it is configurable. To solve the linked issue the configuration would be:

```yaml
redundant_type_annotations:
  ignored_attributes:
    - Parameter
```

This configuration still defaults to only `IBDesignable` for retrocompatibility.

And, while I was at it and as @SimplyDanny suggested in the issue's comments, I took the opportunity to rewrite this rule using SwiftSyntax. Let me know what you think!

tonell-m commented:
Just finished resolving test and linking issues, sorry for the noise

tonell-m commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift:
```diff
@@ -1,9 +1,120 @@
-struct RedundantTypeAnnotationRule: OptInRule, SubstitutionCorrectableRule {
-    var configuration = SeverityConfiguration<Self>(.warning)
+struct RedundantTypeAnnotationRule: SwiftSyntaxCorrectableRule, OptInRule {
+    var configuration = RedundantTypeAnnotationConfiguration()
```
This var can't be private because it's part of the [`Rule`](https://github.com/realm/SwiftLint/blob/6366fb85669c7a3e487fffea80285bafdccb7a8c/Source/SwiftLintCore/Protocols/Rule.swift#L20) protocol

SimplyDanny commented on CHANGELOG.md:
```diff
@@ -15,6 +15,9 @@
   Extensions are an exception; more than one is allowed.  
   [Muhammad Zeeshan](https://github.com/mzeeshanid)
   [#2802](https://github.com/realm/SwiftLint/issues/2802)
+* Rewrite `redundant_type_annotation` rule with SwiftSyntax and add a new `ignored_attributes` option that allows disabling the rule for properties that are marked with at least one of the configured attributes
```
You may the rule's ID and your name to the list in line 19/22 and only mention the new attribute as an enhancement.

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift:
```diff
@@ -1,9 +1,143 @@
 
-struct RedundantTypeAnnotationRule: OptInRule, SubstitutionCorrectableRule {
-    var configuration = SeverityConfiguration<Self>(.warning)
+struct RedundantTypeAnnotationRule: SwiftSyntaxCorrectableRule, OptInRule {
```
Please use `@SwiftSyntaxRule` and remove the explicit `makeVisitor` implementation.

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift:
```diff
@@ -1,9 +1,143 @@
+}
+
+extension RedundantTypeAnnotationRule {
     static let description = RuleDescription(
```
Why is that moved down here? Let's have it right at the beginning after the `configuration` again.

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift:
```diff
@@ -1,9 +1,143 @@
+        // This binding must have both a type declaration and an initializer
+        // sequence for it to be potentially redundant.
+        guard doesNotContainIgnoredAttributes,
+              let binding = bindings.last,
```
Might be fine for now, but actually restricting the rule to the last binding only is not necessary. It could as well check all other bindings.

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift:
```diff
@@ -1,9 +1,143 @@
+
+        // If the initializer is a boolean expression, we consider using the `Bool` type
+        // annotation as redundant.
+        if initializer.as(BooleanLiteralExprSyntax.self) != nil {
```
```suggestion
        if initializer.is(BooleanLiteralExprSyntax.self) {
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift:
```diff
@@ -1,9 +1,143 @@
+                // Parse generic arguments if there are any.
+                var genericArguments = ""
+                if let genericArgumentsClauseBytes = type.genericArguments?.trimmed.syntaxTextBytes {
+                    genericArguments = String(bytes: genericArgumentsClauseBytes, encoding: .utf8) ?? ""
```
The aspect of generic arguments is actually more involved, e.g. `let s: Set<Int> = Set.init([])` should not trigger while `let s: Set = Set<Int>([])` should. An there are more special cases here. We may keep it as simple as it is here, but need to make sure that all OSS findings remain stable with the rewrite.

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift:
```diff
@@ -1,9 +1,143 @@
+            if let calledExpression = functionCall.calledExpression.as(DeclReferenceExprSyntax.self) {
+                // Parse generic arguments if there are any.
+                var genericArguments = ""
+                if let genericArgumentsClauseBytes = type.genericArguments?.trimmed.syntaxTextBytes {
```
Something like `type.genericArguments?.trimmedDescription` might be helpful to avoid the explicit parsing.

SimplyDanny reviewed (changes requested):
Thanks for taking up this rewrite! This is actually a bit more involved than adding the option for ignored attributes only. So I have a few comments ...

garricn commented on Source/SwiftLintBuiltInRules/Rules/RuleConfigurations/RedundantTypeAnnotationConfiguration.swift:
```diff
@@ -0,0 +1,11 @@
+
+    @ConfigurationElement(key: "severity")
+    var severityConfiguration = SeverityConfiguration<Parent>(.warning)
+    @ConfigurationElement(key: "ignored_annotations")
+    var ignoredAnnotations = Set<String>(["IBInspectable"])
```
The word `ignore` might be more appropriate than `ignored` because the configuration element is "configuring the rule to _ignore_ certain annotations". Also, the code base [appears](https://github.com/search?q=repo%3Arealm%2FSwiftLint%20ignore&type=code) to use `ignore`/`ignores` more than `ignored`.

Side note, the PR title and description mention ignore "attributes", not "annotations". I'm just pointing out for consistency purposes.

Great work!

```suggestion
    @ConfigurationElement(key: "ignore_annotations")
    var ignoreAnnotations = Set<String>(["IBInspectable"])
```

tonell-m replied:
You're right it should be attributes and not annotations, it took me a while to figure out which was the actually correct wording for Swift 😅 I'll fix that up!

As for the use of past tense, for me `ignored_attributes` suggests that the given attributes will be ignored by the linter while `ignore_attributes` suggests that this property is a flag that enables ignoring all attributes. Though I agree that for the sake of concistency it is better here to use `ignore` 👍

SimplyDanny replied:
> As for the use of past tense, for me `ignored_attributes` suggests that the given attributes will be ignored by the linter while `ignore_attributes` suggests that this property is a flag that enables ignoring all attributes.

I'm with @tonell-m about this. `ignore_attributes` sounds like a boolean flag.

> Though I agree that for the sake of concistency it is better here to use `ignore` 👍

Can you provide an example where it is really used like that? I can only find flags following this name pattern. `ignored_attributes` seems like a better fit, to be honest.

garricn commented:
I'm really excited about this work! After this merges, perhaps we can also introduce another configuration element named `ignore_literals` which can address the issues raised by this [PR](https://github.com/realm/SwiftLint/pull/4756) as well as the [concerns](https://github.com/realm/SwiftLint/pull/4756#issuecomment-1452585211) of @SimplyDanny therein in terms of not limiting the configuration element to just ignoring booleans.

tonell-m commented:
@SimplyDanny Thanks for the review!

After your comments I've rewrote the rule with a new approach that seems to be more flexible. Instead of visiting `VariableDeclSyntax` nodes I use `PatternBindingSyntax` and `OptionalBindingConditionSyntax` as entry points and moved the logic into a `TypeAnnotationSyntax` extension. That way it will check for all binding patterns even if there are multiple for a single var/let keyword, and it also works the same of guard/if clauses. 
I've also improved the handling of generics to handle more complex cases as you suggested, so `var s: Set = Set<Int>([])` will trigger and not `var s: Set<Int> = Set([])`. I've added a bunch of tests with various configurations to ensure that this is working as expected, let me know if you think of other cases that haven't been handled here :)

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift:
```diff
@@ -72,98 +124,183 @@ struct RedundantTypeAnnotationRule: OptInRule, SubstitutionCorrectableRule {
+private extension RedundantTypeAnnotationRule {
+    final class Visitor: ViolationsSyntaxVisitor<ConfigurationType> {
+        override func visit(_ node: VariableDeclSyntax) -> SyntaxVisitorContinueKind {
+            guard node.doesNotContainIgnoredAttributes(for: configuration) else {
```
This is a good idea. However, it skips all other redundant type annotations nested in the variable declaration, e.g.

```swift
@IgnoreMe
var j: Int {
    let i: Int = Int(1) // will not trigger, but should
    return i
}
```

Therefore it's better to check the parent's parent of `PatternBindingSyntax` for the attribute.

tonell-m replied:
Good call! I've made the update as you suggested

tonell-m commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift:
```diff
@@ -1,8 +1,10 @@
 import Foundation
```
Yes indeed

SimplyDanny commented on CHANGELOG.md:
```diff
@@ -16,17 +16,25 @@
   [Muhammad Zeeshan](https://github.com/mzeeshanid)
   [#2802](https://github.com/realm/SwiftLint/issues/2802)
 
+* Add new `ignore_attributes` option to `redundant_type_annotation` rule
+  that allows disabling the rule for properties that are marked with at least
+  one of the configured attributes
```
The two spaces at the end are important:
```suggestion
* Add new `ignore_attributes` option to `redundant_type_annotation` rule
  that allows disabling the rule for properties that are marked with at least
  one of the configured attributes.  
```

SimplyDanny commented on CHANGELOG.md:
```diff
@@ -16,17 +16,25 @@
   * `identifier_name`
   * `let_var_whitespace`
   * `multiline_literal_brackets`
   * `opening_brace`
   * `void_return`
+  * `redundant_type_annotation`
```
Sorted. 😬
```suggestion
  * `opening_brace`
  * `redundant_type_annotation`
  * `void_return`
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift:
```diff
@@ -28,15 +43,36 @@ struct RedundantTypeAnnotationRule: OptInRule, SubstitutionCorrectableRule {
+            """),
+            Example("@IgnoreMe var a: Int = Int(5)", configuration: ["ignore_attributes": ["IgnoreMe"]]),
+            Example("""
+            var a: Int = {
```
```suggestion
            var a: Int {
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift:
```diff
@@ -52,13 +88,41 @@ struct RedundantTypeAnnotationRule: OptInRule, SubstitutionCorrectableRule {
+            Example("@DontIgnoreMe var a↓: Int = Int(5)", configuration: ["ignore_attributes": ["IgnoreMe"]]),
+            Example("""
+            @IgnoreMe
+            var a: Int = {
```
```suggestion
            var a: Int {
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift:
```diff
@@ -72,98 +136,188 @@ struct RedundantTypeAnnotationRule: OptInRule, SubstitutionCorrectableRule {
+            Example("var num: Int = Int.random(0..<10)"): Example("var num = Int.random(0..<10)"),
+            Example("""
+            @IgnoreMe
+            var a: Int = {
```
```suggestion
            var a: Int {
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift:
```diff
@@ -72,98 +136,188 @@ struct RedundantTypeAnnotationRule: OptInRule, SubstitutionCorrectableRule {
-                $0.1 == [.keyword, .identifier, .typeidentifier, .identifier] ||
-                $0.1 == [.keyword, .identifier, .typeidentifier, .keyword]
+    final class Rewriter: ViolationsSyntaxRewriter {
+        private let configuration: RedundantTypeAnnotationConfiguration
```
```suggestion
        private let configuration: ConfigurationType
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift:
```diff
@@ -72,98 +136,188 @@ struct RedundantTypeAnnotationRule: OptInRule, SubstitutionCorrectableRule {
+    final class Rewriter: ViolationsSyntaxRewriter {
+        private let configuration: RedundantTypeAnnotationConfiguration
+
+        init(configuration: RedundantTypeAnnotationConfiguration,
```
```suggestion
        init(configuration: ConfigurationType,
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/RedundantTypeAnnotationRule.swift:
```diff
@@ -72,98 +136,188 @@ struct RedundantTypeAnnotationRule: OptInRule, SubstitutionCorrectableRule {
-        let substring = file.stringView.substring(with: range)
-        let components = substring.components(separatedBy: "=")
+private extension TypeAnnotationSyntax {
+    func isRedundant(for configuration: RedundantTypeAnnotationConfiguration,
```
`configuration` not needed in here, is it?

tonell-m replied:
Indeed!

SimplyDanny commented:
Looking at the OSS findings, most of them look valid. 

However, [there are cases](https://github.com/wordpress-mobile/WordPress-iOS/blob/12065e659d39e59a7e9a01a45caf31112dee49b8/WordPress/Classes/ViewRelated/Gutenberg/Collapsable%20Header/Collabsable%20Header%20Filter%20Bar/CollabsableHeaderFilterCollectionViewCell.swift#L83:26) where a sequence of `MemberAccessExpression`s leads to the final element/function. An easy example would be `let a: A = A.b.c` They were previously discovered by the rule, so they should also be discoverable by the rewritten rule.

Admittedly, these cases are debatable, but for the sake of a rewrite, the rule should behave as close to the previous version as it can.

@tonell-m: Would you consider these cases in your PR?

SimplyDanny commented on CHANGELOG.md:
```diff
@@ -16,17 +16,25 @@
 
+* Add new `ignore_attributes` option to `redundant_type_annotation` rule
+  that allows disabling the rule for properties that are marked with at least
+  one of the configured attributes.
```
Please append two spaces to the end of the description. They are required to enforce a line break in the rendered Markdown.

tonell-m commented:
> Looking at the OSS findings, most of them look valid.
>
> However, [there are cases](https://github.com/wordpress-mobile/WordPress-iOS/blob/12065e659d39e59a7e9a01a45caf31112dee49b8/WordPress/Classes/ViewRelated/Gutenberg/Collapsable%20Header/Collabsable%20Header%20Filter%20Bar/CollabsableHeaderFilterCollectionViewCell.swift#L83:26) where a sequence of MemberAccessExpressions leads to the final element/function. An easy example would be let a: A = A.b.c They were previously discovered by the rule, so they should also be discoverable by the rewritten rule.
>
> Admittedly, these cases are debatable, but for the sake of a rewrite, the rule should behave as close to the previous version as it can.
>
> @tonell-m: Would you consider these cases in your PR?

@SimplyDanny This case is now handled as well. Do you think we also need to handle cases where there is a mix of `MemberAccessExprSyntax` and `FunctionCallExprSyntax`? Like `let a: A = A.f().b`

SimplyDanny commented:
> > Looking at the OSS findings, most of them look valid.
> > However, [there are cases](https://github.com/wordpress-mobile/WordPress-iOS/blob/12065e659d39e59a7e9a01a45caf31112dee49b8/WordPress/Classes/ViewRelated/Gutenberg/Collapsable%20Header/Collabsable%20Header%20Filter%20Bar/CollabsableHeaderFilterCollectionViewCell.swift#L83:26) where a sequence of MemberAccessExpressions leads to the final element/function. An easy example would be let a: A = A.b.c They were previously discovered by the rule, so they should also be discoverable by the rewritten rule.
> > Admittedly, these cases are debatable, but for the sake of a rewrite, the rule should behave as close to the previous version as it can.
> > @tonell-m: Would you consider these cases in your PR?
> 
> @SimplyDanny This case is now handled as well. Do you think we also need to handle cases where there is a mix of `MemberAccessExprSyntax` and `FunctionCallExprSyntax`? Like `let a: A = A.f().b`

That depends on what the previous implementation did. I guess, it triggered, hence the rewritten rule should trigger too.

SimplyDanny commented:
> That depends on what the previous implementation did. I guess, it triggered, hence the rewritten rule should trigger too.

Rebased and addressed this point.
